commit 0a76a8c765c7b8327c59646284fa78c3c27e5490	0a76a8c765c7b8327c59646284fa78c3c27e5490
Author: erouault <erouault>
Date:   Wed Jan 11 16:13:50 2017 +0000

    * libtiff/tif_jpeg.c: validate BitsPerSample in JPEGSetupEncode() to avoid
    undefined behaviour caused by invalid shift exponent.
    Fixes http://bugzilla.maptools.org/show_bug.cgi?id=2648

diff --git a/ChangeLog b/ChangeLog
index 65176404..8e202a2c 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,9 @@
+2017-01-11 Even Rouault <even.rouault at spatialys.com>
+
+	* libtiff/tif_jpeg.c: validate BitsPerSample in JPEGSetupEncode() to avoid
+	undefined behaviour caused by invalid shift exponent.
+	Fixes http://bugzilla.maptools.org/show_bug.cgi?id=2648
+
 2017-01-11 Even Rouault <even.rouault at spatialys.com>
 
 	* libtiff/tif_dir.c, tif_dirread.c, tif_dirwrite.c: implement various clampings
diff --git a/libtiff/tif_jpeg.c b/libtiff/tif_jpeg.c
index 6c17c388..192989a9 100644
--- a/libtiff/tif_jpeg.c
+++ b/libtiff/tif_jpeg.c
@@ -1632,6 +1632,13 @@ JPEGSetupEncode(TIFF* tif)
                             "Invalig horizontal/vertical sampling value");
                     return (0);
                 }
+                if( td->td_bitspersample > 16 )
+                {
+                    TIFFErrorExt(tif->tif_clientdata, module,
+                                 "BitsPerSample %d not allowed for JPEG",
+                                 td->td_bitspersample);
+                    return (0);
+                }
 
 		/*
 		 * A ReferenceBlackWhite field *must* be present since the
